libvast/include/vast/system/exporter.hpp

```diff
@@ -84,7 +84,9 @@ struct exporter_state {
 /// @param options The query options.
 /// @param pipelines The applied pipelines.
 exporter_actor::behavior_type
-exporter(exporter_actor::stateful_pointer<exporter_state> self, expression expr,
-         query_options options, std::vector<pipeline>&& pipelines);
+exporter(exporter_actor::stateful_pointer<exporter_state> self,
+         std::optional<expression> expr,
+         std::vector<std::string> spawn_arguments, query_options options,
```

Comment:
Passing in both doesn't make much sense to me. The conversion from `spawn_arguments` to `expression` should likely take place in the `spawn_exporter` and `spawn_counter` functions respectively, such that we only need to normalize and validate the query that we got passed in.

Reply:
We're trying to get it out of `spawn_arguments`, which is populated in `spawn_command`. The more specific `spawn_{exporter,counter}` do argument checking/parsing and then spawn the actual component. To summarize:
- Use `spawn_{exporter,counter}` to parse the expression
- Validate the expression in the exporter/counter directly

This ensures that all requirements are directly encoded in the types of the function arguments.

Reply:
Okay, I got it now. I addressed all your concerns in https://github.com/tenzir/vast/pull/2850/commits/3f38bb3f70f35d2f442084f444575aee3f903942.